### PR TITLE
Feature 78028:   UI General info tab

### PR DESCRIPTION
**Pull request by janhagevold**

### CODE REVIEW

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/index.tsx [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            <HeaderContainer>
+                <Typography variant="h6">Participants</Typography>
+            </HeaderContainer>
+            <ProjectInfoContainer>
+                <ProjectInfoDetail>
+                    <Typography variant="caption">Meeting point</Typography>
+                    <Typography variant="body1">{generalInfo.meetingPoint}</Typography>
+                </ProjectInfoDetail>
+            </ProjectInfoContainer>
+            <ParticipantsTable participants={participantList} completePunchOut={completePunchOut} />
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/index.tsx:88]: Is it possible to make the table horizontally scrollable if it cannot fit on the screen, or is this difficult when working with widths? From what I see this is not a big issue (and I think I recall this being tricky), so dont spend too much time on it.

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+                </SpinnerContainer>
+            )}
+            <CustomTable>
+                <Head>
+                    <Row>
+                        <Cell as="th" scope="col" style={{verticalAlign: 'middle'}}>Attendance list </Cell>
+                        <Cell as="th" scope="col" style={{verticalAlign: 'middle'}}>Representative </Cell>
+                        <Cell as="th" scope="col" style={{verticalAlign: 'middle'}}>Outlook response</Cell>
+                        <Cell as="th" scope="col" style={{verticalAlign: 'middle'}}>Attended</Cell>
+                        <Cell as="th" scope="col" style={{verticalAlign: 'middle'}}>Note/Comment</Cell>
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx:92]: Just being picky, but I think we should decide on either note or comment. You can ask Jon. If he said to have it like this then it is fine

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+                        <Cell as="th" scope="col" style={{verticalAlign: 'middle'}}>Signed at</Cell>
+                    </Row>
+                </Head>
+                <Body>
+                    {data && data.length > 0 && data.map((participant: Participant, index: number) => (
+                        <Row key={index} as="tr">
+                            <Cell as="td" style={{verticalAlign: 'middle'}}>{participant.role}</Cell>
+                            <Cell as="td" style={{verticalAlign: 'middle'}}>{participant.name}</Cell>
+                            <Cell as="td" style={{verticalAlign: 'middle'}}>{participant.response}</Cell>
+                            <Cell as="td" style={{verticalAlign: 'middle', minWidth: '160px'}}>
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx:103]: Just a note - I have asked Jon as well - but I think the attended and note properties should be disabled until the user clicks "Complete punch out", then he will get a popup (https://statoildeveloper.visualstudio.com/Statoil.NextGenProCoSys/_backlogs/backlog/Statoil.NextGenProCoSys%20Team/Features/?workitem=77740) where he will do this, and then complete. We will leave it like this for now.

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    const [data, setData] = useState<Participant[]>([]);
+    const [loading, setLoading] = useState<boolean>(false);
+    const btnRef = useRef<HTMLButtonElement>();
+
+    useEffect(() => {
+        setData(participants);
+    }, [participants]);
+
+    const getSignedProperty = (id: string | undefined, signedBy: string | undefined, handleCompletePunchOut: (index: number) => void): JSX.Element => {
+        // TODO: Determine participant permission for current user
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx:31]: More detail in another comment, but I think signed will be "-" for everyone except maybe first two participants. Then we will have getCompleted and getAccepted instead, which will probably be properties on the invitation.

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+                <CustomTooltip title={tooltipText} arrow>
+                    <Button tooltip={tooltipText} ref={btnRef} onClick={handleCompletePunchOut}>Complete punch out</Button>
+                </CustomTooltip>
+            );
+        } else {
+            return <span>-</span>;
+        }
+    };
+
+    const handleCompletePunchOut = async (index: number): Promise<any> => {
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx:45]: We will probably change this a bit as well (to show popup). 
As mentioned in pevious comments, I dont think we will set signedAt/By, but complete the punchout, and get new response from endpoint where completedAt/By is set.

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomCells.tsx [hunk 6]
```diff
@@ -0,0 +1,34 @@
+import { CustomSwitch, CustomTextField } from './CustomFields';
+
+import React from 'react';
+import { Typography } from '@material-ui/core';
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomCells.tsx:4]: Use Typography from EDS

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomCells.tsx [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+interface AttendedProps {
+    status: boolean;
+    onChange: (e: any) => void;
+}
+
+export const AttendedEditCell = ({status, onChange}: AttendedProps): JSX.Element => {
+    return (
+        <div style={{display: 'flex', alignItems: 'center'}}>
+            <CustomSwitch checked={status} onChange={(): void => onChange(!status)}/>
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomCells.tsx:14]: Check if EDS has a swith we can use 
https://eds-storybook-react-development.azurewebsites.net/?path=/story/components-selection-controls-switch--switch-control

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomFields.tsx [hunk 8]
```diff
@@ -0,0 +1,33 @@
+import { Switch, TextField, Tooltip, withStyles } from '@material-ui/core';
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomFields.tsx:1]: EDS has text field and Switch.
The Tooltip in EDS is buggy, so keep using the one from material-ui :-)

**Review by bratli — CHANGES_REQUESTED** [review]

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx (hunk 2 above)
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx:92]: Column renamed Notes

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx (hunk 4 above)
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx:31]: To be determined in another task

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx (hunk 5 above)
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx:45]: Adjusted in popup task

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomFields.tsx (hunk 8 above)
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomFields.tsx]: done

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomCells.tsx (hunk 7 above)
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomCells.tsx]: done

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomCells.tsx (hunk 6 above)
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomCells.tsx]: done

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/index.tsx [hunk 9]
```diff
@@ -0,0 +1,94 @@
+import { Container, DateTimeItem, DetailContainer, HeaderContainer, ProjectInfoContainer, ProjectInfoDetail } from './style';
+import React, { useEffect, useState } from 'react';
+import { getFormatDate, getFormatTime } from './utils';
+
+import { Participant } from './types';
+import ParticipantsTable from './ParticipantsTable';
+import { Typography } from '@material-ui/core';
```
[the remark this answers is not in the document]
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/index.tsx]: done

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/index.tsx [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+                        </DateTimeItem>
+                    </DetailContainer>
+                </ProjectInfoDetail>
+            </ProjectInfoContainer>
+            <HeaderContainer>
+                <Typography variant="h6">Participants</Typography>
+            </HeaderContainer>
+            <ProjectInfoContainer>
+                <ProjectInfoDetail>
+                    <Typography variant="caption">Meeting point</Typography>
```
[the remark this answers is not in the document]
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/index.tsx]: location used

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/index.tsx [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            } catch (error) {
+                reject(error);
+            }
+        });
+    };
+
+    return (
+        <Container>
+            <HeaderContainer>
+                <Typography variant="h6">General generalInfo</Typography>
```
[the remark this answers is not in the document]
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/index.tsx]: fixed

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts [hunk 12]
```diff
@@ -0,0 +1,26 @@
+export type ResponseType = 'Attending' | 'Tentative' | 'Not responded' |'Declined';
+
+export type Participant = {
+    id?: string;
```
[the remark this answers is not in the document]
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts]: done

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+    type: string;
+    title: string;
+    description: string;
+    punchRoundTime: {
+        from: Date;
+        to: Date;
+    },
+    participants: Participant[];
+    meetingPoint: string;
+    invitationSent: boolean;
```
[the remark this answers is not in the document]
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts]: removed invitation

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    project: string;
+    type: string;
+    title: string;
+    description: string;
+    punchRoundTime: {
+        from: Date;
+        to: Date;
+    },
+    participants: Participant[];
+    meetingPoint: string;
```
[the remark this answers is not in the document]
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts]: location used

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts [hunk 15]
```diff
@@ -0,0 +1,26 @@
+export type ResponseType = 'Attending' | 'Tentative' | 'Not responded' |'Declined';
+
+export type Participant = {
+    id?: string;
+    name: string;
+    role: string;
+    response: ResponseType;
+    attended: boolean;
+    notes: string;
+    signedBy?: string;
```
[the remark this answers is not in the document]
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts]: CompetedAt, CompletedBy, AcceptedBy, AcceptedAt columns added to Invitation (GeneralInfoType)

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/dummyData.ts [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    },
+    participants: [
+        {
+            id: '0',
+            name: 'Anne ASPH',
+            role: 'Contractor',
+            response: 'Attending',
+            attended: true,
+            notes: '',
+            signedBy: undefined,
```
[the remark this answers is not in the document]
Comment (janhagevold) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/dummyData.ts]: done

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (bratli) [reply src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts]: We can fix this when getting data from endpoint (unless you have something else to fix), but id here is a int.

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/index.tsx [hunk 17]
```diff
@@ -0,0 +1,93 @@
+import { CompletedType, Participant } from './types';
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/index.tsx:1]: This in not used - will it be?

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+    },
+    participants: Participant[];
+    location: string;
+    completedAt?: Date;
+    completedBy?: string;
+    acceptedAt?: Date;
+    acceptedBy?: string;
+}
+
+export type CompletedType = {
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/types.d.ts:30]: Not used, will it be?

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomCells.tsx [hunk 19]
```diff
@@ -0,0 +1,42 @@
+import { Switch, TextField } from '@equinor/eds-core-react';
+import { Tooltip, withStyles } from '@material-ui/core';
+
+import React from 'react';
+import { Typography } from '@equinor/eds-core-react';
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/CustomCells.tsx:5]: move in with top import from EDS

**Review by bratli — APPROVED** [review]
See comments

File: src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx [hunk 20]
```diff
@@ -103,14 +100,11 @@ const ParticipantsTable = ({participants, completePunchOut}: Props): JSX.Element
...
                             </Cell>
-                            <Cell as="td" style={{verticalAlign: 'middle', width: '40%'}}>
+                            <Cell as="td" style={{verticalAlign: 'middle', width: '40%', minWidth: '200px'}}>
                                 <NotesEditCell value={participant.notes} onChange={handleEditNotes} index={index} />
                             </Cell>
                             <Cell as="td" style={{verticalAlign: 'middle', minWidth: '160px'}}>
-                                {getSignedProperty(participant.id, participant.signedBy, () => handleCompletePunchOut(index))}
-                            </Cell>
-                            <Cell as="td" style={{verticalAlign: 'middle', minWidth: '150px'}}>
-                                {participant.signedAt ? `${getFormatDate(participant.signedAt)} ${getFormatTime(participant.signedAt)}` : '-'}
```
Comment (bratli) [inline src/modules/InvitationForPunchOut/views/ViewIPO/GeneralInfo/ParticipantsTable/index.tsx]: I think we still need signet at column in table btw!

**Review by bratli — CHANGES_REQUESTED** [review]

**Review by bratli — APPROVED** [review]

**Pull request merged**